File: src/main/java/com/budgetbuddy/service/DataCleanupService.java
package com.budgetbuddy.service;

import com.budgetbuddy.model.User;
import com.budgetbuddy.repository.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class DataCleanupService {

    private static final Logger logger = LoggerFactory.getLogger(DataCleanupService.class);

    @Autowired
    private TransactionRepository transactionRepository;

    @Autowired
    private CategoryOverspendingAlertRepository categoryOverspendingAlertRepository;

    @Autowired
    private MoneyLeakRepository moneyLeakRepository;

    @Autowired
    private SavingsProjectionRepository savingsProjectionRepository;

    @Autowired
    private WeekendOverspendingRepository weekendOverspendingRepository;

    @Autowired
    private SpendingPatternRepository spendingPatternRepository;

    /**
     * Clear all transaction data and financial guidance data for a user
     * WARNING: This will permanently delete all transactions and all financial analysis data
     */
    @Transactional
    public void clearAllTransactionAndGuidanceData(User user) {
        logger.warn("⚠️ Starting data cleanup for user: {} (ID: {})", user.getName(), user.getId());
        
        try {
            Long userId = user.getId();
            
            // 1. Clear all financial guidance data first (to avoid foreign key issues)
            clearFinancialGuidanceData(user);
            
            // 2. Delete all transactions for this user
            List<com.budgetbuddy.model.Transaction> userTransactions = transactionRepository.findAll()
                .stream()
                .filter(t -> t.getUser().getId().equals(userId))
                .collect(Collectors.toList());
            
            if (!userTransactions.isEmpty()) {
                transactionRepository.deleteAll(userTransactions);
                logger.info("✅ Deleted {} transactions for user {}", userTransactions.size(), userId);
            } else {
                logger.info("ℹ️ No transactions found for user {}", userId);
            }
            
            logger.info("✅ Successfully cleared all transaction and financial guidance data for user: {} (ID: {})", 
                user.getName(), userId);
        } catch (Exception e) {
            logger.error("❌ Error clearing transaction and guidance data for user {}: {}", 
                user.getId(), e.getMessage(), e);
            throw new RuntimeException("Failed to clear data: " + e.getMessage(), e);
        }
    }

    /**
     * Clear all financial guidance data for a user
     */
    @Transactional
    private void clearFinancialGuidanceData(User user) {
        Long userId = user.getId();
        logger.info("Clearing financial guidance data for user: {} (ID: {})", user.getName(), userId);
        
        try {
            // Delete Category Overspending Alerts
            List<com.budgetbuddy.model.CategoryOverspendingAlert> categoryAlerts = 
                categoryOverspendingAlertRepository.findByUserAndIsActiveTrue(user);
            if (!categoryAlerts.isEmpty()) {
                categoryOverspendingAlertRepository.deleteAll(categoryAlerts);
                logger.info("✅ Deleted {} category overspending alerts", categoryAlerts.size());
            }
            
            // Delete Money Leaks
            List<com.budgetbuddy.model.MoneyLeak> moneyLeaks = 
                moneyLeakRepository.findByUserAndIsActiveTrueOrderByAnnualAmountDesc(user);
            if (!moneyLeaks.isEmpty()) {
                moneyLeakRepository.deleteAll(moneyLeaks);
                logger.info("✅ Deleted {} money leaks", moneyLeaks.size());
            }
            
            // Delete Savings Projections
            List<com.budgetbuddy.model.SavingsProjection> savingsProjections = 
                savingsProjectionRepository.findByUserAndYearOrderByProjectionDateDesc(user, null);
            // Get all savings projections for the user
            List<com.budgetbuddy.model.SavingsProjection> allSavingsProjections = 
                savingsProjectionRepository.findAll()
                    .stream()
                    .filter(sp -> sp.getUser().getId().equals(userId))
                    .collect(Collectors.toList());
            if (!allSavingsProjections.isEmpty()) {
                savingsProjectionRepository.deleteAll(allSavingsProjections);
                logger.info("✅ Deleted {} savings projections", allSavingsProjections.size());
            }
            
            // Delete Weekend Overspending
            List<com.budgetbuddy.model.WeekendOverspending> weekendOverspending = 
                weekendOverspendingRepository.findByUserAndIsActiveTrue(user);
            if (!weekendOverspending.isEmpty()) {
                weekendOverspendingRepository.deleteAll(weekendOverspending);
                logger.info("✅ Deleted {} weekend overspending records", weekendOverspending.size());
            }
            
            // Delete Spending Patterns
            List<com.budgetbuddy.model.SpendingPattern> spendingPatterns = 
                spendingPatternRepository.findByUserAndIsActiveTrue(user);
            if (!spendingPatterns.isEmpty()) {
                spendingPatternRepository.deleteAll(spendingPatterns);
                logger.info("✅ Deleted {} spending patterns", spendingPatterns.size());
            }
            
            logger.info("✅ Successfully cleared all financial guidance data for user: {} (ID: {})", 
                user.getName(), userId);
        } catch (Exception e) {
            logger.error("❌ Error clearing financial guidance data for user {}: {}", 
                userId, e.getMessage(), e);
            throw e;
        }
    }
}

File: mobile-version/app/src/main/java/com/budgetbuddy/mobile/service/DataCleanupService.kt
package com.budgetbuddy.mobile.service

import com.budgetbuddy.mobile.data.dao.*
import android.util.Log
import kotlinx.coroutines.flow.first

/**
 * Data Cleanup Service - Equivalent to Spring Boot's DataCleanupService
 * Clears all transaction data and financial guidance data for a user
 */
class DataCleanupService(
    private val transactionDao: TransactionDao,
    private val categoryOverspendingAlertDao: CategoryOverspendingAlertDao,
    private val moneyLeakDao: MoneyLeakDao,
    private val savingsProjectionDao: SavingsProjectionDao,
    private val weekendOverspendingDao: WeekendOverspendingDao,
    private val spendingPatternDao: SpendingPatternDao,
    private val spendingPredictionDao: SpendingPredictionDao,
    private val financialNudgeDao: FinancialNudgeDao
) {
    
    /**
     * Clear all transaction data and financial guidance data for a user
     * WARNING: This will permanently delete all transactions and all financial analysis data
     * Equivalent to Spring Boot's clearAllTransactionAndGuidanceData()
     */
    suspend fun clearAllTransactionAndGuidanceData(userId: Long) {
        Log.w("DataCleanupService", "⚠️ Starting data cleanup for user: $userId")
        
        try {
            // 1. Clear all financial guidance data first (to avoid foreign key issues)
            clearFinancialGuidanceData(userId)
            
            // 2. Delete all transactions for this user
            transactionDao.deleteAllTransactionsForUser(userId)
            Log.i("DataCleanupService", "✅ Deleted all transactions for user $userId")
            
            Log.i("DataCleanupService", "✅ Successfully cleared all transaction and financial guidance data for user: $userId")
        } catch (e: Exception) {
            Log.e("DataCleanupService", "❌ Error clearing transaction and guidance data for user $userId: ${e.message}", e)
            throw RuntimeException("Failed to clear data: ${e.message}", e)
        }
    }
    
    /**
     * Clear all financial guidance data for a user
     * Equivalent to Spring Boot's clearFinancialGuidanceData()
     */
    private suspend fun clearFinancialGuidanceData(userId: Long) {
        Log.i("DataCleanupService", "Clearing financial guidance data for user: $userId")
        
        try {
            // Delete Category Overspending Alerts
            categoryOverspendingAlertDao.deleteByUserId(userId)
            Log.i("DataCleanupService", "✅ Deleted category overspending alerts")
            
            // Delete Money Leaks
            moneyLeakDao.deleteByUserId(userId)
            Log.i("DataCleanupService", "✅ Deleted money leaks")
            
            // Delete Savings Projections
            savingsProjectionDao.deleteByUserId(userId)
            Log.i("DataCleanupService", "✅ Deleted savings projections")
            
            // Delete Weekend Overspending
            weekendOverspendingDao.deleteByUserId(userId)
            Log.i("DataCleanupService", "✅ Deleted weekend overspending records")
            
            // Delete Spending Patterns
            try {
                spendingPatternDao.deactivateAllPatternsForUser(userId)
                // Also delete all patterns
                val allPatterns = spendingPatternDao.getAllPatternsByUser(userId).first()
                allPatterns.forEach { pattern ->
                    spendingPatternDao.deletePattern(pattern)
                }
                Log.i("DataCleanupService", "✅ Deleted spending patterns")
            } catch (e: Exception) {
                Log.w("DataCleanupService", "Could not delete spending patterns: ${e.message}")
            }
            
            // Delete Spending Predictions
            try {
                spendingPredictionDao.deleteAllPredictionsForUser(userId)
                Log.i("DataCleanupService", "✅ Deleted spending predictions")
            } catch (e: Exception) {
                Log.w("DataCleanupService", "Could not delete spending predictions: ${e.message}")
            }
            
            // Delete Financial Nudges
            try {
                val allNudges = financialNudgeDao.getAllNudgesByUser(userId).first()
                allNudges.forEach { nudge ->
                    financialNudgeDao.deleteNudge(nudge)
                }
                Log.i("DataCleanupService", "✅ Deleted financial nudges")
            } catch (e: Exception) {
                Log.w("DataCleanupService", "Could not delete financial nudges: ${e.message}")
            }
            
            Log.i("DataCleanupService", "✅ Successfully cleared all financial guidance data for user: $userId")
        } catch (e: Exception) {
            Log.e("DataCleanupService", "❌ Error clearing financial guidance data for user $userId: ${e.message}", e)
            throw e
        }
    }
}

